### PR TITLE
feat(code-block): left-align tabs, fix valid children issue

### PR DESCRIPTION
**Pull request by zchsh**
🎟️ [Asana Task - Alignment](https://app.asana.com/0/1100423001970639/1201866381269498/f)
🎟️ [Asana Task - Children validation fix](https://app.asana.com/0/1100423001970639/1201881376116200/f)
🔍 [Preview Link](https://react-components-git-zscode-tabs-alignment-hashicorp.vercel.app/components/codeblock)

---

## Description

This PR...

This PR also fixes a bug in `CodeTabs` related to validation of children elements, which prevented some use in JSX. With this fix in place, `CodeTabs` should correctly validate the use of `CodeBlock` children in JSX.

### PR Checklist 🚀

Items in this checklist may not may not apply to your PR, but please consider each item carefully.

- [x] Add Asana and Preview links above.
- [x] Conduct thorough self-review.
- [x] Add or update tests as appropriate.
- [x] Conduct reasonable cross browser testing for both compatibility and responsive behavior (We have a [Sauce Labs](https://app.saucelabs.com/) account for this, if you don't have access, just ask!).
- [x] Conduct reasonable accessibility review (use the [WAS](https://accessible.org/Web-Accessibility-Standards-WAS-2.pdf) as a guide or an [axe browser plugin](https://www.deque.com/axe/) until we establish more formal checks).
- [x] Identify (in the description above) and document (add Asana tasks on [this board](https://app.asana.com/0/1100423001970639/list)) any technical debt that you're aware of, but are not addressing as part of this PR.

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1201881376116200
  - https://app.asana.com/0/0/1201866381269498

### CODE REVIEW

File: packages/code-block/docs.mdx [hunk 1]
```diff
@@ -275,18 +275,42 @@ function hello() {
 
 The `CodeTabs` component, which is available in Learn and Docs MDX contexts, allows authors to create multi-snippet tabbed code blocks. `<CodeTabs>` is a wrapper component around child code blocks. Each child must be either a fenced code block, or must be a `<CodeBlockConfig>` which wraps a fenced code block.
 
+By default, the tab options are shown at the left of the component. When space is limited, tabs will automatically collapse into a dropdown menu. When a `heading` is provided to `CodeTabs`, the heading will be shown at left, and the tabs (or dropdown menu) will be shown at right.
+
+#### `CodeTabs` Playground
```
Comment (zchsh) [inline packages/code-block/docs.mdx:280]: Added a section in Swingset to help demonstrate the new alignment feature.

File: packages/code-block/docs.mdx [hunk 2]
```diff
@@ -275,18 +275,42 @@ function hello() {
...
+<span class="token punctuation">}</span>\`}
+/>
+
+</CodeTabs>
+`}</ LiveComponent>
+
 #### When to use `CodeTabs`
 
-Tabs should only be used when switching between similar content types such as:
+Tabs should only be used when switching between equivalent approaches to a particular task or workflow. Common examples include alternate configuration type, programming, languages, or operating systems:
```
Comment (zchsh) [inline packages/code-block/docs.mdx:308]: Rephrased this a bit to get to the point and then show examples - it felt like the previous text I'd written was a bit unclear.

File: packages/code-block/partials/code-tabs/index.js [hunk 3]
```diff
@@ -23,14 +25,31 @@ function CodeTabs({ children, heading, className, tabs, theme = 'dark' }) {
...
-    // For function components in JSX, ie CodeBlock and CodeBlockConfig,
-    // tabChild.type is a function whose name we need
-    if (typeof tabChild.type == 'function') type = tabChild.type.name
-    // In MDX contexts, the component type is captured in mdxType
-    if (typeof tabChild.props.mdxType === 'string')
+    // For JSX primitives, the type is captured by the type property
+    if (typeof tabChild.type == 'string' || typeof tabChild.type == 'number') {
+      type = tabChild.type
+      // For function components, accept CodeBlock or CodeBlockConfig.
+    } else if (typeof tabChild.type === 'function') {
```
Comment (zchsh) [inline packages/code-block/partials/code-tabs/index.js:32]: This builds on the approach Bryce suggested to maintain the existing behaviour of flagging all invalid types in a an error message. Due to the minification concern, there's a bit of repetition here to make sure the intended names make it into the error message; my thinking is that's a decent tradeoff for having clearer, easier-to-debug error messages.

Have also made a small style change to use braces for the `if` blocks 👍

File: packages/code-block/partials/code-tabs/index.test.js [hunk 4]
```diff
@@ -1,14 +1,15 @@
...
+import CodeBlockConfig from '../code-block-config'
 import CodeTabsProvider from '../../provider'
 
 // getAllByText is needed, as we sometimes render dummy elements
 // before the main, displayed element, eg for overflow detection
 // and for simpler copy-to-clipboard logic
 function getSecondByText(text) {
   /* eslint-disable-next-line no-unused-vars */
-  const [firstElem, secondElem] = screen.getAllByText(text)
+  const [_firstElem, secondElem] = screen.getAllByText(text)
```
Comment (zchsh) [inline packages/code-block/partials/code-tabs/index.test.js:12]: Nit, this is an unused variable, prefixing with `_` makes VSCode happy (at least in my case).

File: packages/code-block/partials/code-tabs/index.test.js [hunk 5]
```diff
@@ -25,6 +26,35 @@ it('should accept a className prop, and place it on the root element', () => {
   expect(container.firstChild.className).toContain(customClassName)
 })
 
+it('should accept CodeBlockConfig or pre elements as children', () => {
```
Comment (zchsh) [inline packages/code-block/partials/code-tabs/index.test.js:29]: Added a test to assert that the changes made don't affect the existing behaviour.

File: packages/code-block/partials/code-tabs/partials/tabs-as-tabs/index.js [hunk 6]
```diff
@@ -2,16 +2,20 @@ import React from 'react'
 import classNames from 'classnames'
 import s from './style.module.css'
 
-function TabsAsTabs({ tabLabels, activeTabIdx, setActiveTabIdx }) {
+function TabsAsTabs({ tabLabels, activeTabIdx, setActiveTabIdx, hasHeading }) {
   return (
-    <div className={s.tabsAsTabs}>
+    <div className={classNames(s.tabsAsTabs, { [s.hasHeading]: hasHeading })}>
```
Comment (zchsh) [inline packages/code-block/partials/code-tabs/partials/tabs-as-tabs/index.js:7]: To achieve left-alignment when a `heading` is not present, but maintain the existing right-alignment when a `heading` is present, we add some styling for `TabsAsTabs` (and the space-limited alternate `TabsAsDropdowns`) which is applied conditionally based on the presence of a `heading`.

**Comment by zchsh** [comment]
@ashleemboyer This fixes the `CodeTabs` issue (encountered on the Install views), and updates tabs to be left-aligned (unless a `heading` is provided). I've proven this out in https://github.com/hashicorp/dev-portal/pull/169. Let me know what you think! 🙇‍♂️

File: packages/code-block/partials/code-tabs/index.test.js (hunk 4 above)
Comment (ashleemboyer) [reply packages/code-block/partials/code-tabs/index.test.js:12]: If you want I think you could update this to not name any variable at all:

```suggestion
  const [, secondElem] = screen.getAllByText(text)
```

Comment (zchsh) [reply packages/code-block/partials/code-tabs/index.test.js:12]: nice, TIL!

**Review by ashleemboyer — APPROVED** [review]
LGTM! 🥇👏🏻

**Comment by zchsh** [comment]
Wrestling with the canary release workflow, it seems to be hanging for some reason (usually takes < 1 min and taking WAY longer). Have run into this before, there's no reason why the code change in that suggestion should affect it, I think it's just flaky. Trying one more time, but going to merge even if it fails (cause it's some strange issue with the workflow, not with that latest commit).

**Pull request merged**